ExPlat: Self-Contained Client API

jessie-ross opened the pull request:
**This PR implements and unit tests the client API:**
- Includes the SSR safety mechanism and related tests.

Part of a series rebuilding the reference WIP PR #48475

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run `yarn test` within `/packages/explat-client`

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

jessie-ross committed:
Add ExPlat SAC package

jessie-ross committed:
Implement and test internals

jessie-ross committed:
Improve type imports

jessie-ross committed:
Update types

jessie-ross committed:
Document state functions

jessie-ross committed:
Minor name fix

jessie-ross committed:
Spell EA out

jessie-ross committed:
Use devDeps for calypso-polyfills

jessie-ross committed:
Test requests

jessie-ross committed:
Minor fix

jessie-ross committed:
Improve internals

jessie-ross committed:
Implement the client

jessie-ross committed:
Fix names

yanirs commented on packages/explat-client/src/create-explat-client.ts:
```diff
@@ -0,0 +1,199 @@
+
+				return fetchedExperimentAssignment;
+			} catch ( e ) {
+				// We need to be careful about errors thrown in the catch block too:
```
All this nesting of try-catch blocks isn't great... It seems like the main concern is that `logError()` would throw an error. To work around it, how about we add a local function like this:

```typescript
const safeLogError = ( message ) => {
  try {
    config.logError( { message, experimentName } );
  } catch ( e )  {
    if ( config.isDevelopmentMode ) {
      throw e;
    }
  }
}
```

Then, instead of the nested try-catch blocks we can do:

```typescript
  ...
  return fetchedExperimentAssignment;
} catch ( e ) {
  safeLogError( e.message );
  if ( config.isDevelopmentMode ) {
     throw e;
  }
}
// Fetching failed and we're not in development mode.
try {
  // ... fallback 1 code (lines 134-148)
} catch ( e ) {
  // ... fallback 2 code, with safeLogError()
}
```

Also, do we even need the second try-catch? Lines 134-148 shouldn't throw exceptions, should they?

jessie-ross replied:
My idea here is to make this failsafe, and the second try-catch makes that simpler: if we have any code here we should assume it can go wrong and have a backup plan. If we push any failures into the backup plan we then only need to ensure the backup plan itself can't go wrong rather than the rest of the code.

In this case the backup plan is the final `createFallbackExperimentAssignment`.


> Lines 134-148 shouldn't throw exceptions, should they?

They shouldn't but now we are covered in the case they do.

Also it is not just `logError` but also `storeExperimentAssignment` - the internals can also throw.

yanirs replied:
OK, it seems a bit too defensive to me, but I still think that flattening the try-catches is worthwhile. Having three nested levels of try-catch blocks with error names like `e`, `e2`, and `e3` is harder to follow than the flow I suggested, which makes it clear that we only get into fallbacks outside dev mode.

yanirs commented on packages/explat-client/src/create-ssr-safe-mock-explat-client.ts:
```diff
@@ -0,0 +1,25 @@
+import { ExPlatClient } from './create-explat-client';
+import { createFallbackExperimentAssignment } from './internal/experiment-assignments';
+
+export default function createSsrSafeMockExPlatClient( config: Config ): ExPlatClient {
```
How about we call this `createSsrSafeDummyExPlatClient()`? There is no mocking in the test-mock sense.

Also, it'd be good to document this, and I think it can live in the `create-explat-client.ts` file.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,940 @@
+const mockedGetAnonId = jest.fn();
+const mockedLogError = jest.fn();
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore; Not using the full Config
```
These lines should be removed now that we're using the full config.

```suggestion
```

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,940 @@
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+global.window = {};
```
It'd be more self-documenting to add a couple of helpers `setBrowserContext()` and `setSsrContext()` that'd change the `window` value instead of the three lines to override the window in multiple places. They can also be combined into a single helper like:

```typescript
function setBrowserContext( browserContext: boolean ) {
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore
	global.window = browserContext ? {} : undefined;
}
```

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,940 @@
+	` );
+	} );
+
+	// eslint-disable-next-line jest/expect-expect
```
I think it'd be a good idea to expect something instead of suppressing the `expect-expect` warning. For example, we can check that the return value of each `createExPlatClient()` call isn't null.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,940 @@
+const mockedLogError = jest.fn();
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore; Not using the full Config
+const mockedConfig: Config = {
```
What's the reasoning behind sharing a global mock between tests rather than creating a unique mock for each test? It feels a bit brittle, and I'm also unsure why we're naming the mocked functions as separate consts.

Instead, we can do something like this: 
```typescript
function createConfigMock( overrides: Partial<Config> ): Config {
	return {
		logError: jest.fn(),
		fetchExperimentAssignment: jest.fn(),
		getAnonId: jest.fn(),
		isDevelopmentMethod: false,
		...overrides
	}
}
```

With this, we can do stuff like `configMock = createConfigMock()` and  `configMock.getAnonId.mockImplementationOnce(...)`, or pass in the specific mock overrides to the creation function.

jessie-ross replied:
Ah I did this to make `mockFetchExperimentAssignmentToMatchExperimentAssignment` easier to use but fair enough, not too hard to make it work.

yanirs replied:
Yeah, I was thinking we could also make the overrides more sophisticated so that passing an `ExperimentAssignment` object as the value of `fetchExperimentAssignment` would also create the mock, but I'm not sure it'd be that clean without playing with it.

jessie-ross replied:
So I have added `createConfigMock` so far but now I am thinking that this doesn't add much - mocks are reset anyway in-between tests, and making this work well with typescript is a bit of a pain as it needs to fit the two different uses - as a Config and as a mocked Config (e.g. to access `.mockImplement()`).

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,940 @@
+					[ experimentAssignment.experimentName ]: experimentAssignment.variationName,
+				},
+			},
+			1
```
As noted on another PR, these `delayedValue()` calls are hard to reason about and should at least be documented, though constants would be better than magic numbers.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,940 @@
+		await expect(
+			client.loadExperimentAssignment( validExperimentAssignment.experimentName )
+		).resolves.toEqual( validExperimentAssignment );
+		expect( allMockedConfigCalls() ).toMatchInlineSnapshot( `
+		Object {
+		  "fetchExperimentAssignment": Array [
+		    Array [
+		      Object {
+		        "anonId": undefined,
+		        "experimentName": "experiment_name_a",
+		      },
+		    ],
+		  ],
+		  "getAnonId": Array [
+		    Array [],
+		  ],
+		  "logError": Array [],
+		}
```
Why do we care what calls the client made as long as it returns what we expect based on the mocks? These sort of checks may make the implementation too rigid.

It seems like the only mock that we should generally check is `logError()`, since it's a side-effect that we normally care about.

jessie-ross replied:
To me this is integration testing, ensuring the boundary of the client stays constant.

yanirs replied:
I'm not sure that this counts as the boundary of the client, as the exact internal calls aren't a part of the contract. In many cases, the fact that the calls happened is implied by the observed outputs.

The main output that we don't observe is that of `logError()`, which is worth checking when we expect errors.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,940 @@
+		await expect(
+			client.loadExperimentAssignment( 'the-invalid-experiment-name' )
+		).rejects.toThrow( 'Invalid experimentName: the-invalid-experiment-name' );
+		expect( allMockedConfigCalls() ).toMatchInlineSnapshot( `
+		Object {
+		  "fetchExperimentAssignment": Array [],
+		  "getAnonId": Array [],
+		  "logError": Array [
+		    Array [
+		      Object {
+		        "experimentName": "the-invalid-experiment-name",
+		        "message": "Invalid experimentName: the-invalid-experiment-name",
+		      },
+		    ],
+		  ],
+		}
+	` );
```
This is a good example of a case where we really shouldn't care about the mocks called. Eliminating such cases would make this big test file much easier to digest and maintain.

I'll hold off on reading the rest until we've discussed the points made so far.

jessie-ross replied:
If the size of the file is a concern I can split it by the top level describes.

yanirs replied:
It's just a lot to review &ndash; I don't think we need to split it to files, just ensure that the tests only test what they need to test. Having less code to read/maintain is a nice side effect. :slightly_smiling_face: 

My point still stands: In this case, we don't care about the mocks since an error was thrown, so I think we should remove the mock check. This applies to other cases where the behaviour is enough to tell us that the mocks were used as expected.

We can get a second opinion on this from people who work in testing, though.

jessie-ross committed:
Merge branch 'trunk' into add/explat-sac-internals-request

jessie-ross committed:
Use Promise.resolve

jessie-ross committed:
Improve requests and related tests

- Ensure ttl is greater than 0
- Use constansts for 0 and 1 delays
- Minor cleanup

jessie-ross committed:
Convert state to use a class

jessie-ross committed:
Minor cleanup

jessie-ross committed:
Use experimentAssignmentStore

jessie-ross committed:
Simplify error flow

jessie-ross committed:
Minor cleanup

jessie-ross committed:
Move dummy explat client in with the real one

jessie-ross committed:
Fix tests

jessie-ross committed:
Minor cleanup

jessie-ross committed:
Only snapshot logError calls

jessie-ross committed:
Minor cleanup

jessie-ross committed:
Use createConfigMock

jessie-ross committed:
Use constants for delay

jessie-ross committed:
Minor cleanup

yanirs commented on packages/explat-client/src/create-explat-client.ts:
```diff
@@ -0,0 +1,214 @@
+				} catch ( e2 ) {}
+				if ( config.isDevelopmentMode ) {
+					throw e;
+				}
```
This should be removed as it's unreachable (we're guaranteed not to be in development mode at this point).

yanirs commented on packages/explat-client/src/create-explat-client.ts:
```diff
@@ -0,0 +1,214 @@
+					return storedExperimentAssignment;
+				}
+
+				// We are syncronously trying to retreive and then store a fallback which means this fallback will
```
`retreive` -> `retrieve`

yanirs commented on packages/explat-client/src/create-explat-client.ts:
```diff
@@ -0,0 +1,214 @@
+					config.logError( {
+						message: e.message,
+						experimentName,
+						isSecondaryError: 'true',
```
Is the `isSecondaryError` value going to be read by anything? If not, let's just change the message, e.g., the message here can be `'Fallback Error: ' + e.message`.

jessie-ross replied:
Yeah it ends up LogStashed, I think it is useful to diagnose errors in the wild.

yanirs replied:
Maybe we can make it more descriptive then? It looks odd that we're passing `'true'` as a string, but I see that's needed from `logError`'s signature. We can add something like `context: 'failed fallback attempt'` and `context: 'failed fetch attempt'` for the initial try.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,697 @@
+	` );
+		jest.useRealTimers();
+	} );
+	it( `logError throws/secondary error: should attempt to log secondary error and return fallback`, async () => {
```
If I understand the test case description correctly, there should be a secondary error logged, right?

yanirs commented on packages/explat-client/src/create-explat-client.ts:
```diff
@@ -0,0 +1,214 @@
+						message: e.message,
+						experimentName,
+					} );
+				} catch ( e2 ) {}
```
As [suggested previously](https://github.com/Automattic/wp-calypso/pull/49832#discussion_r572594842), I think that adding a thin wrapper around `logError` (aka `safeLogError`) would make this much more readable.

Alternatively, giving the exceptions more informative names would probably help. Something like:

* First `e`: `initialFetchError`
* Second `e`: `fallbackError`
* First and second `e2`: `ignoredLoggingError`

yanirs commented on packages/explat-client/src/test/index.ts:
```diff
@@ -0,0 +1,39 @@
+	it( 'should return the real client when run in a browser context', () => {
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		global.window = {};
```
Should use the `setBrowserContext()` and `setSsrContext()` helpers in this file too.

yanirs commented on packages/explat-client/src/test/index.ts:
```diff
@@ -0,0 +1,39 @@
+		// @ts-ignore
+		global.window = {};
+		const exPlatIndex = require( '../index' );
+		expect( exPlatIndex.createExPlatClient.toString() ).toBe(
+			createBrowserExPlatClient.toString()
+		);
```
Wouldn't this work without the `toString()` calls?

jessie-ross replied:
No it doesn't seem to, and this seemed to work well enough for our purposes although I can understand it isn't ideal.

jessie-ross replied:
<img width="588" alt="Screen Shot 2021-02-18 at 4 00 02 pm" src="https://user-images.githubusercontent.com/971886/108324030-5ebedd80-7202-11eb-8df7-c4e675ee8f71.png">

yanirs replied:
Ah, I guess it may be because of the way imports work... Anyway, I agreed that using `toString()` is good enough.

yanirs commented on packages/explat-client/src/test/create-ssr-safe-mock-explat-client.ts:
```diff
@@ -0,0 +1,86 @@
+const mockedGetAnonId = jest.fn();
+const mockedLogError = jest.fn();
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore; Not using the full Config
```
These lines are unnecessary now.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,697 @@
+
+const spiedMonotonicNow = jest.spyOn( Timing, 'monotonicNow' );
+
+const mockedFetchExperimentAssignment = jest.fn();
+const mockedGetAnonId = jest.fn();
+const mockedLogError = jest.fn();
+const createMockedConfig = ( isDevelopmentMode = false ): Config => ( {
+	logError: mockedLogError,
+	fetchExperimentAssignment: mockedFetchExperimentAssignment,
+	getAnonId: mockedGetAnonId,
+	isDevelopmentMode,
+} );
+const DEVELOPMENT_MODE = true;
```
Continuing the discussion from https://github.com/Automattic/wp-calypso/pull/49832#discussion_r572610336, this sharing of globals across test cases is still problematic, and the current setup is confusing. When looking at an isolated test case, it's not immediately obvious that a name like `mockedGetAnonId` is related to the mock config we just created, and declaring `DEVELOPMENT_MODE = true` at the top of the file sounds like it's setting development mode to true.

Can you please explain further why the suggestion I made won't work? Copying here for context:

```typescript
function createConfigMock( overrides: Partial<Config> ): Config {
	return {
		logError: jest.fn(),
		fetchExperimentAssignment: jest.fn(),
		getAnonId: jest.fn(),
		isDevelopmentMethod: false,
		...overrides
	}
}
```

With this, we can do stuff like `configMock = createConfigMock()` and  `configMock.getAnonId.mockImplementationOnce(...)`, or pass in the specific mock overrides to the creation function.

If TypeScript is unhappy with some of the mocking, we can suppress specific warnings for the file / function.

I can also play with it if you prefer.

yanirs replied:
Ah! On further reflection, I think that a part of my confusion here is that technically, we're not mocking the config. We are using Jest's mocking helpers to create a valid config, which is passed to the `createExPlatClient()` function. In general, clients should be free to put whatever they want in the config, as long as it adheres to the interface.

This is just an aside, though. I still think that reducing the reliance on globals would make the tests clearer.

jessie-ross replied:
I have changed it closer to what you suggested and it didn't end up being so bad.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,697 @@
+		);
+	} );
+
+	it( `shouldn't call any deps if only initialized`, () => {
```
Shouldn't we check all the dependencies to match the test description?

We can also drop this test case completely, as I don't think it's very interesting &ndash; we get plenty of coverage from the other tests.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,697 @@
+	} );
+	it( `[anonId] should successfully load an ExperimentAssignment`, async () => {
+		mockedGetAnonId.mockImplementationOnce( () => delayedValue( 'the-anon-id-123', ZERO_DELAY ) );
+		mockFetchExperimentAssignmentToMatchExperimentAssignment( validExperimentAssignment );
```
For this test to match our expectations, we should set up the mock so that it only returns the experiment assignment if the anon ID is `the-anon-id-123`.

If we make the mocking changes suggested above, it'd be something like `mockFetchExperimentAssignmentToMatchExperimentAssignment( { experimentAssignment: validExperimentAssignment, expectedAnonId: 'the-anon-id-123' } )`

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,697 @@
+} );
+const DEVELOPMENT_MODE = true;
+
+function mockFetchExperimentAssignmentToMatchExperimentAssignment(
+	experimentAssignment: ExperimentAssignment
+) {
+	spiedMonotonicNow.mockImplementationOnce( () => experimentAssignment.retrievedTimestamp );
+	mockedFetchExperimentAssignment.mockImplementationOnce( () =>
```
This mock function doesn't match the signature of the function we're mocking. We should accept the `experimentName` and `anonId` parameters and check their value. The formatting is a mess, but I think that something like this would work:

```typescript
function mockFetchExperimentAssignmentToMatchExperimentAssignment(
	{ experimentAssignment, expectedAnonId } :
	{ experimentAssignment: ExperimentAssignment; expectedAnonId: string | null; }
) {
	spiedMonotonicNow.mockImplementationOnce( () => experimentAssignment.retrievedTimestamp );
	mockedFetchExperimentAssignment.mockImplementationOnce( ( experimentName, anonId ) => {
		expect( experimentName ).toBe( experimentAssignment.experimentName );
		expect( anonId ).toBe( expectedAnonId );
		...
```

See also below for how it's useful in the context of specific tests.

jessie-ross replied:
I have just ended up checking the `mockedConfig.requestExperimentAssignment.mock.calls`.

yanirs replied:
Why? How is it better than using a more realistic mock?

jessie-ross replied:
<img width="1355" alt="Screen Shot 2021-02-18 at 4 58 17 pm" src="https://user-images.githubusercontent.com/971886/108332226-86b23f00-720a-11eb-9fd6-17c6e38f85ca.png">

It gets a bit complicated as jest works by throwing errors and the error is occurring within a config function. I am also sometimes returning a different experimentAssignment to what was requested to check other things.

I could still add more checks on the requestExperimentAssignment calls, maybe snapshotting them?

yanirs replied:
Ah, never mind then. I think what we have is good enough.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,697 @@
+		await expect(
+			client.loadExperimentAssignment( 'the-invalid-experiment-name' )
+		).resolves.toEqual( {
+			experimentName: 'fallback_experiment_assignment',
```
As ntoed on #49829, I think there's no harm in returning the invalid name here. It's less surprising than `fallback_experiment_assignment`.

jessie-ross replied:
See the comment there - it is not a valid ExperimentAssignment and would fail various run-time checks.

jessie-ross replied:
I have replied there and have also changed the name to indicate that it is an invalid name.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,697 @@
+	` );
+	} );
+
+	it( `[developmentMode] Could not fetch ExperimentAssignment: should throw`, async () => {
```
Not critical, but this file may be easier to follow if we split the development mode tests to nested `describe`s. That is:

```typescript
describe( 'ExPlatClient.loadExperimentAssignment single-use', () => {
	describe( 'in production mode, () => {
		// ... production mode tests ...
	} );

	describe( 'in development mode, () => {
		// ... development mode tests ...
	} );
} );
```

jessie-ross replied:
I was thinking about this but I like that the development and production tests are next to each other - it is easier to compare behaviour, makes refactoring a bit more obvious, and makes an omissions of a corresponding developer test clear.

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,697 @@
+		expect( mockedLogError.mock.calls ).toMatchInlineSnapshot( `Array []` );
+	} );
+
+	it( `[developmentMode] should respect the ttl`, async () => {
```
This is almost identical to the previous test. Is the behaviour supposed to be the same? It'd be good to reduce duplication here to make the intention clearer.

jessie-ross committed:
Update packages/explat-client/src/internal/test/requests.ts

Co-authored-by: Yanir Seroussi <yanir.seroussi@automattic.com>

jessie-ross committed:
Update packages/explat-client/src/internal/test/requests.ts

Co-authored-by: Yanir Seroussi <yanir.seroussi@automattic.com>

yanirs commented on packages/explat-client/src/test/create-explat-client.ts:
```diff
@@ -0,0 +1,697 @@
+		const experimentAssignmentD = client.loadExperimentAssignment(
+			validExperimentAssignment.experimentName
+		);
+		let errorA;
+		try {
+			await experimentAssignmentA;
+		} catch ( e ) {
+			errorA = e;
+		}
```
It'd be nice to add an `awaitException()` helper that'd return the error. Then we can do `const errorA = awaitException( experimentAssignmentA )`. The helper can also fail if the exception isn't thrown.

jessie-ross committed:
Merge remote-tracking branch 'upstream/add/explat-sac-internals-request' into add/explat-sac-api

jessie-ross committed:
Merge remote-tracking branch 'upstream/add/explat-sac-internals-request' into fix/explat-sac-internals

jessie-ross committed:
Minor style fixes

jessie-ross committed:
Add originalExperimentName

jessie-ross committed:
Minor cleanup

jessie-ross committed:
Merge branch 'fix/explat-sac-internals' into add/explat-sac-api

jessie-ross committed:
Minor fix

jessie-ross committed:
Minor cleanup

jessie-ross committed:
Improve logging

jessie-ross committed:
Fix spelling

jessie-ross committed:
Minor cleanup

jessie-ross committed:
Change originalExperimentName to requested

jessie-ross committed:
Remove regex test on isName

jessie-ross committed:
Clean up types

jessie-ross committed:
Merge branch 'fix/explat-sac-internals' into add/explat-sac-api

jessie-ross committed:
Make mockedConfig non-global

jessie-ross committed:
Fix tests for invalidExperiment name

jessie-ross committed:
Fix fallbackError test

jessie-ross committed:
Clean up mockedConfig in SSR tests

jessie-ross committed:
Check anonId

jessie-ross committed:
Dedupe the repect ttl test

yanirs commented on packages/explat-client/src/test/index.ts:
```diff
@@ -0,0 +1,47 @@
+	jest.resetModules();
+} );
+
+function setBrowserContext() {
+	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+	// @ts-ignore
+	global.window = {};
+}
+
+function setSsrContext() {
+	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+	// @ts-ignore
+	global.window = undefined;
+}
```
It'd be good to put these functions in the common utility file rather than duplicate them.

yanirs reviewed (approved):
I still think it's worth addressing https://github.com/Automattic/wp-calypso/pull/49832#discussion_r578219835 and reducing duplication a bit, but should be good to go along with the rest of the stack (and code style fixes). :ship:

jessie-ross committed:
Move setXContext to test-common

jessie-ross committed:
Merge remote-tracking branch 'upstream/trunk' into add/explat-sac-api
